src/utils/component-loader.ts

```diff
@@ -0,0 +1,165 @@
+import * as path from 'path'
+import * as fs from 'fs'
+import { OverridableComponent } from '../frontend/utils/overridable-component'
+import { relativeFilePathResolver } from './file-resolver'
+import { ConfigurationError } from '../backend'
+
+export interface ComponentDetails {
+  overrides: boolean
+  filePath: string
+}
+
+export default class ComponentLoader {
+  protected components: Record<string, ComponentDetails>
+
+  public constructor(protected readonly themeId?: string) {
+    this.components = {}
+  }
+
```

Comment:
I think we will need a way to merge multiple loaders for example when using features like `@adminjs/passwords`.

How about we add something like:

```
public mergeLoader(loader: ComponentLoader) {
   if (this.getThemeId() || loader.getThemeId()) {
     throw new Error('Cannot merge Theme loaders');
   }

   this.components = {
       ...loader.getComponents(),
       ...this.getComponents(),
   }
}
```

Reply:
Feature packages would also export their loaders which you'd have to merge with you own loader.

Reply:
I don't like that approach, I'd rather share a single loader if it's possible with the current API. I'll check the code and see if I can come up with something.